pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -19,32 +19,297 @@
+    """
+    # First we get a set of real eigenvectors
+    _, p = math.linalg.eigh(math.real(uuT) + math.imag(uuT))
+
+    # Next, we will extract the eigenvalues by doing np.dot(uuT, p) / p, and
+    # taking the values from the columns. However, to account that there might
+    # be 0 values in p, we go column by column and extract the eigenvalues from
+    # the non-zero elements only.
+    ev_p = []
+
+    uuT_p = math.dot(uuT, math.cast_like(p, 1j))
+
+    for col_idx in range(4):
+        non_zero_index = np.nonzero(np.logical_not(np.isclose(p[:, col_idx], np.zeros(4))))
+        non_zero_uuT_p = math.take(uuT_p[:, col_idx], non_zero_index)[0, 0]
+        non_zero_p = math.take(p[:, col_idx], non_zero_index)[0, 0]
+        ev_p.append(non_zero_uuT_p / math.cast_like(non_zero_p, 1j))
+
+    ev_p = math.cast_like(math.convert_like(ev_p, uuT), 1j)
+
+    return ev_p, p
+
+
+def _extract_su2su2_prefactors(U, V):
+    r"""This function is used for the case of 2 CNOTs and 3 CNOTs. It does something
+    similar as the 1-CNOT case, but there is no special form for one of the
+    SO(4) operations.
+
+    Suppose U, V are SU(4) matrices for which there exists A, B, C, D such that
+    (A \otimes B) V (C \otimes D) = U. The problem is to find A, B, C, D in SU(2)
+    in an analytic and fully differentiable manner.
+
+    This decomposition is possible when U and V are in the same double coset of
+    SU(4), meaning there exists G, H in SO(4) s.t. G (Edag V E) H = (Edag U
+    E). This is guaranteed here by how V was constructed in both the
+    _decomposition_2_cnots and _decomposition_3_cnots methods.
+
+    Then, we can use the fact that E SO(4) Edag gives us something in SU(2) x
+    SU(2) to give A, B, C, D.
+    """
+
+    # A lot of the work here happens in the magic basis. Essentially, we
+    # don't look explicitly at some U = G V H, but rather at
+    #     E^\dagger U E = G E^\dagger V E H
+    # so that we can recover
+    #     U = (E G E^\dagger) V (E H E^\dagger) = (A \otimes B) V (C \otimes D).
+    # There is some math in the paper explaining how when we define U in this way,
+    # we can simultaneously diagonalize functions of U and V to ensure they are
+    # in the same coset and recover the decomposition.
+
+    u = math.dot(math.cast_like(Edag, V), math.dot(U, math.cast_like(E, V)))
+    v = math.dot(math.cast_like(Edag, V), math.dot(V, math.cast_like(E, V)))
+
+    uuT = math.dot(u, math.T(u))
+    vvT = math.dot(v, math.T(v))
+
+    # Get the p and q in SO(4) that diagonalize uuT and vvT respectively (and
+    # their eigenvalues)
+    ev_p, p = _extract_so4_eigensystem(uuT)
+    ev_q, q = _extract_so4_eigensystem(vvT)
+
+    # Now we must sort the eigenvalues into the same order if this is not
+    # a special case where it is already done.
+    if not math.allclose(ev_p, ev_q):
+        new_q_order = []
+        for _, eigval in enumerate(ev_p):
+            are_close = [math.allclose(x, eigval) for x in ev_q]
```

Comment:
I am not sure why these 3 lines are not showing as covered; the 2- and 3-CNOT test cases should all be going through this portion.